File: skills/code-architect/references/code-examples/ruby/strategy.rb
# frozen_string_literal: true

# Strategy — a family of interchangeable algorithms behind one interface.

# Strategies: each exposes #apply(items) and can be swapped freely.
class AscendingOrder
  def apply(items) = items.sort
end

class DescendingOrder
  def apply(items) = items.sort.reverse
end

# A lambda is a perfectly valid strategy too (duck typing).
ByLength = ->(items) { items.sort_by(&:length) }

# Context holds a strategy and delegates the varying step to it.
class Sorter
  def initialize(strategy) = @strategy = strategy
  attr_writer :strategy

  def run(items)
    @strategy.respond_to?(:apply) ? @strategy.apply(items) : @strategy.call(items)
  end
end

if __FILE__ == $PROGRAM_NAME
  data = %w[pear fig apple kiwi]
  sorter = Sorter.new(AscendingOrder.new)

  puts sorter.run(data).inspect

  sorter.strategy = DescendingOrder.new
  puts sorter.run(data).inspect

  sorter.strategy = ByLength
  puts sorter.run(data).inspect
end

File: skills/code-architect/references/code-examples/typescript/Strategy.ts
// Strategy — a family of interchangeable algorithms behind one interface.

interface SortStrategy {
  sort(input: number[]): number[];
}

class AscendingStrategy implements SortStrategy {
  sort(input: number[]): number[] {
    return [...input].sort((a, b) => a - b);
  }
}

class DescendingStrategy implements SortStrategy {
  sort(input: number[]): number[] {
    return [...input].sort((a, b) => b - a);
  }
}

// Context delegates the algorithm choice to a strategy.
class Sorter {
  constructor(private strategy: SortStrategy) {}

  setStrategy(strategy: SortStrategy): void {
    this.strategy = strategy;
  }

  run(input: number[]): number[] {
    return this.strategy.sort(input);
  }
}

function demo(): void {
  const data = [3, 1, 4, 1, 5];
  const sorter = new Sorter(new AscendingStrategy());
  console.log(`asc:  ${sorter.run(data).join(",")}`);
  sorter.setStrategy(new DescendingStrategy());
  console.log(`desc: ${sorter.run(data).join(",")}`);
}

demo();
